core/src/main/scala/org/apache/spark/sql/SnappyParser.scala

```diff
@@ -348,6 +348,42 @@ class SnappyParser(session: SnappySession)
       (d, s.asInstanceOf[Option[Duration]]))
   }
 
+  private def extractGroupingSet(
+      child: LogicalPlan,
+      aggregations: Seq[NamedExpression],
+      groupByExprs: Seq[Expression],
+      groupingSets: Seq[Seq[Expression]]): GroupingSets = {
+    val keyMap = groupByExprs.zipWithIndex.toMap
+    val bitmasks: Seq[Int] = groupingSets.map(set => set.foldLeft(0)((bitmap, col) => {
+      require(keyMap.contains(col), s"$col doesn't show up in the GROUP BY list")
+      bitmap | 1 << keyMap(col)
+    }))
+    GroupingSets(bitmasks, groupByExprs, child, aggregations)
+  }
+
+  protected final def groupingSetExpr: Rule1[Seq[Expression]] = rule {
+    '(' ~ ws ~ (expression + (',' ~ ws)) ~ ws ~ ')'  ~>  ((gs: Seq[Expression]) => gs) |
+        (expression + (',' ~ ws))  ~>  ((gs: Seq[Expression]) => gs) |
+        '(' ~ ws ~ ')'  ~>  (() => (Seq[Expression]()))
+  }
+
+  protected final def cubeRollUpGroupingSet: Rule1[(Seq[Seq[Expression]], Option[String])] = rule {
+    CUBE ~> (() => (Seq(Seq[Expression]()), Option("CUBE"))) |
+    ROLLUP ~> (() => (Seq(Seq[Expression]()),Option("ROLLUP"))) |
+    GROUPING ~ SETS ~ ('(' ~ ws ~ (groupingSetExpr ~ ws).+(',' ~ ws) ~ ws ~ ')' ~ ws)  ~>
```

Comment:
The second and fourth "~ ws" should be removed once the convention I mentioned above is followed. So you can be:
(groupingSetExpr + (',' ~ ws)) or groupingSetExpr.+(',' ~ ws) which is a bit consistent and cleaner

Perhaps we should refactor to add a separate def for (',' ~ ws) which is being used in many places.